contracts/external/cover/CoverBroker.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.18;
+
+import "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-v4/security/ReentrancyGuard.sol";
+
+import "../../interfaces/ICover.sol";
+import "../../interfaces/ICoverBroker.sol";
+import "../../interfaces/IMemberRoles.sol";
+
+/// @dev Allows cover distribution by buying cover in behalf of the caller
+contract CoverBroker is ICoverBroker, ReentrancyGuard {
+
+  // Immutables
+  address owner;
+  ICover cover;
+  IMemberRoles memberRoles;
+
+  // Constants
+  address public constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+  modifier onlyOwner() {
+    if (msg.sender != owner) {
+      revert OnlyOwner();
+    }
+    _;
+  }
+
+  constructor(address _owner, address _cover, address _memberRoles) {
+    owner = _owner;
+    cover = ICover(_cover);
+    memberRoles = IMemberRoles(_memberRoles);
+  }
+
+  /// @dev buys cover in behalf of the caller
+  function buyCover(
+    BuyCoverParams memory params,
+    PoolAllocationRequest[] calldata poolAllocationRequests
+  ) external payable nonReentrant returns (uint coverId) {
+    uint ethBalanceBefore = address(this).balance;
+
+    // set the cover owner to the caller
+    params.owner = msg.sender;
+
+    // call cover.buyCover with msg.value and params
+    coverId = cover.buyCover{value: msg.value}(params, poolAllocationRequests);
+
+    if (address(this).balance > ethBalanceBefore) {
```

Comment:
Seems like only ETH cover buys are possible. AFAIR this is going to be the case for the first use-case but we should implement non-ETH cover buys as well.

Reply:
add ERC20 payment support